Cython/Compiler/PyrexTypes.py

```diff
@@ -4577,7 +4580,8 @@ def is_promotion(src_type, dst_type):
             return src_type.is_float and src_type.rank <= dst_type.rank
     return False
 
-def best_match(arg_types, functions, pos=None, env=None, args=None):
+def best_match(arg_types, functions, pos=None, env=None, args=None,
+               validate_types_fully=False):
```

Comment:
Hmm. Do we need this flag? I'd rather not add a "do your job" and a "really do your job" kind of behavioural distinction. Comparing argument types doesn't seem like a bad idea in general.

Reply:
I've removed it... we'll see if everything else still works.

Reply:
I came to the conclusion that the flag is needed unfortunately. I changed the name to `validate_types_for_operators` to make it seem a little less like "do your job" and a "really do your job".

Reply:
It still feels like it's working by accident. If things only work because of the shortcut, in cases where we have exactly one candidate, then the whole machinery that gets triggered by having two (or more) candidates starts to feel fragile.

I looked at the test results and yes, removing the shortcut breaks it pretty badly. But that suggests to me even stronger that the shortcut is a general issue. At least its position at the front. Maybe we should rather have it as a fallback in the end. And maybe keep a tighter variant in the beginning only for the one _really_, really obvious case.

Reply:
I'll have another go at it in a bit - I agree it does seem very fragile.

I think the issue for the functions that broke was that they had a generic PyObject, but required a `long`. That's obviously something that we can handle (it's a conversion with an exception if a long isn't suitable), but it doesn't fit into the "pick the best C++ overload" logic. I suspect it probably should fit into the logic, but with low priority and only when there's one option.